### PR TITLE
POD-575: Update to support new Terra adapter workflows

### PR DESCRIPTION
**Pull request by gordonwade**
### Purpose
This PR updates several functions that are used to generate JSON metadata conforming to HCA schemas (see: https://schema.humancellatlas.org/) in order to support ingest of analyses run in Terra. Several additional functions are added in order to generate new types of metadata to meet the new requirements.

### Changes

* New functions added to `optimus.py` and `smartseq2.py` to support creation of tsv files from analysis metadata
* Updates to several functions in `create_analysis_metadata.py` to support generation of new UUIDs, updates to `analysis_protocol` and `analysis_process` files
* Addition of `create_file_descriptor.py` to create `file_descriptor` metadata
* Addition of `create_links.py` to create `links` metadata
* Addition of `create_reference_file.py` to create `reference_file` metadata
* Updates to existing tests and addition of new tests for all new files
* Addition of test metadata for the new tests
* Addition of new entry points for the new files
* Dockerfile updated to allow calculation of crc32c hash

### Review Instructions

These changes work as expected for the staging of post-analysis data from Terra. **Please note that this PR contains breaking changes for the updated functions**

### CODE REVIEW

File: pipeline_tools/shared/submission/create_analysis_metadata.py [hunk 1]
```diff
@@ -3,23 +3,30 @@
...
 import re
+import uuid
 import arrow
-from pipeline_tools.shared.submission.format_map import EXTENSION_TO_FORMAT
+from pipeline_tools.shared.submission.format_map import (
+    EXTENSION_TO_FORMAT,
+    NAMESPACE,
+    get_uuid5,
+    convert_datetime,
+)
```
Comment (rexwangcc) [inline pipeline_tools/shared/submission/create_analysis_metadata.py:18]: great formatting!!

File: pipeline_tools/shared/submission/create_analysis_metadata.py [hunk 2]
```diff
@@ -113,6 +119,16 @@ def create_analysis_protocol(
         'computational_method': method,
         'type': get_analysis_protocol_type(),
     }
+
+    string_to_hash = json.dumps(analysis_protocol, sort_keys=True)
+    entity_id = str(uuid.uuid5(NAMESPACE, string_to_hash)).lower()
```
Comment (rexwangcc) [inline pipeline_tools/shared/submission/create_analysis_metadata.py:124]: Just curious, why use UUID5 here not UUID4? Is this a new spec for the entity IDs?

File: pipeline_tools/shared/submission/create_file_descriptor.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        'crc32c': crc32c,
+        'file_id': file_id,
+        'file_version': file_version,
+        'file_name': get_relative_file_location(file_path),
+    }
+
+    return file_descriptor
+
+
+def get_datetime_from_file_info(file_info):
```
Comment (rexwangcc) [inline pipeline_tools/shared/submission/create_file_descriptor.py:53]: Is this function being used anywhere?

**Review by rexwangcc — APPROVED** [review]
Looks great to me! There are a lot of changes and really appreciate your work on this! I left a few comments but in general LGTM

File: pipeline_tools/shared/submission/create_analysis_metadata.py (hunk 2 above)
Comment (gordonwade) [reply pipeline_tools/shared/submission/create_analysis_metadata.py:124]: Exactly right - in the new spec, the UUIDs need to be deterministic.

File: pipeline_tools/shared/submission/create_file_descriptor.py (hunk 3 above)
Comment (gordonwade) [reply pipeline_tools/shared/submission/create_file_descriptor.py:53]: Good point! I think this got dropped during one of the iterations - removing now.

**Comment by gordonwade** [comment]
Thanks for the feedback, @rexwangcc! Made a few updates, accordingly.

File: pipeline_tools/pipelines/optimus/optimus.py [hunk 4]
```diff
@@ -185,3 +188,25 @@ def create_optimus_input_tsv(uuid, version, dss_url):
         f.write(f"{chemistry}")
 
     print('Finished writing files')
+
+
+def create_optimus_inputs_tsv_from_analysis_metadata(metadata_file):
```
Comment (samanehsan) [inline pipeline_tools/pipelines/optimus/optimus.py:193]: Does this metadata file contain the Cromwell metadata? A docstring here would help!

**Review by rexwangcc — APPROVED** [review]
LGTM! (We may want to turn off the Jenkins Integration tests since they are not testing anything really..)

File: pipeline_tools/shared/submission/format_map.py [hunk 5]
```diff
@@ -19,3 +22,13 @@
     "[.]npz$": "npz",
     "[.]npy$": "npy",
 }
+
+NAMESPACE = uuid.UUID('c6591d1d-27bc-4c94-bd54-1b51f8a2456c')
```
Comment (samanehsan) [inline pipeline_tools/shared/submission/format_map.py:26]: Just curious but where does this uuid come from?

File: pipeline_tools/tests/shared/submission/test_create_file_descriptor.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+def data_file():
+    def _data_file(file_name):
+        return (
+            f'{Path(os.path.split(__file__)[0]).absolute().parents[1]}/data/{file_name}'
+        )
+
+    return _data_file
+
+
+class TestCreateAnalysisMetadata(object):
```
Comment (samanehsan) [inline pipeline_tools/tests/shared/submission/test_create_file_descriptor.py:32]: Did you mean `TestCreateFileDescriptor` here?

**Review by samanehsan — APPROVED** [review]
This looks great!! I just left a few small comments :octocat:

File: pipeline_tools/shared/submission/format_map.py (hunk 5 above)
Comment (gordonwade) [reply pipeline_tools/shared/submission/format_map.py:26]: Good question - this is just an arbitrary value. We need one to use consistently in order for the UUIDs to be deterministic.

File: pipeline_tools/tests/shared/submission/test_create_file_descriptor.py (hunk 6 above)
Comment (gordonwade) [reply pipeline_tools/tests/shared/submission/test_create_file_descriptor.py:32]: Absolutely. Good catch!

File: pipeline_tools/pipelines/optimus/optimus.py (hunk 4 above)
Comment (gordonwade) [reply pipeline_tools/pipelines/optimus/optimus.py:193]: Another good question - updating the docstring to make it clear that this is Cromwell metadata.

**Pull request merged**